nomad/server.go

```diff
@@ -1193,62 +1193,41 @@ func (s *Server) setupRPC(tlsWrap tlsutil.RegionWrapper) error {
 	return nil
 }
 
-// setupRpcServer is used to populate an RPC server with endpoints. This gets
-// called at startup but also once for every new RPC connection so that RPC
-// handlers can have per-connection context.
-func (s *Server) setupRpcServer(server *rpc.Server, ctx *RPCContext) error {
-
+// setupStreamingEndpoints is used to populate an RPC server with streaming
+// endpoints. This only gets called at server startup.
+func (s *Server) setupStreamingEndpoints(server *rpc.Server, ctx *RPCContext) {
 	// The endpoints are client RPCs and don't include a connection
 	// context. They also need to be registered as streaming endpoints in their
 	// register() methods.
-	//
-	// Streaming RPCs should only be registered once, so before we register
-	// them, we make a check if they haven't been already registered.
 
 	clientAllocs := NewClientAllocationsEndpoint(s)
+	clientAllocs.register()
+	_ = server.Register(clientAllocs)
+
 	fsEndpoint := NewFileSystemEndpoint(s)
+	fsEndpoint.register()
+	_ = server.Register(fsEndpoint)
+
 	agentEndpoint := NewAgentEndpoint(s)
+	agentEndpoint.register()
+	_ = server.Register(agentEndpoint)
 
 	// Event is a streaming-only endpoint so we don't want to register it as a
 	// normal RPC
 	eventEndpoint := NewEventEndpoint(s)
+	eventEndpoint.register()
 
 	// Operator takes a RPC context but also has a streaming RPC that needs to
 	// be registered
 	operatorEndpoint := NewOperatorEndpoint(s, ctx)
-
-	// streamingEndpoint is a helper interface to make working with endpoint
-	// structs easier
-	type streamingEndpoint interface {
-		register()
-	}
-
-	streaming_endpoints := map[string]streamingEndpoint{
-		"Allocations.Exec": clientAllocs,
-		"Agent.Monitor":    agentEndpoint,
-		// fsEndpoint registers "FileSystem.Logs" and "FileSystem.Stream", but
-		// if at least one of them is registered we know the method's been
-		// called
-		"FileSystem.Logs": fsEndpoint,
-		"Event.Stream":    eventEndpoint,
-		// operatorEndpoint registers "Operator.SnapshotSave" and
-		// "Operator.SnapshotRestore"
-		"Operator.SnapshotSave": operatorEndpoint,
-	}
-
-	// If we haven't registered the streaming endpoints before, register them
-	// now.
-	for method, endpoint := range streaming_endpoints {
-		if _, err := s.streamingRpcs.GetHandler(method); err != nil {
-			endpoint.register()
-		}
-	}
-
-	_ = server.Register(clientAllocs)
-	_ = server.Register(fsEndpoint)
-	_ = server.Register(agentEndpoint)
+	operatorEndpoint.register()
 	_ = server.Register(NewOperatorEndpoint(s, ctx))
```

Comment:
The client allocs, FS, agent, and operator endpoints all have non-streaming RPCs as well, so we should move these `_ = server.Register(..)` method calls into `setupRpcServer`. It looks like the event endpoint is the only one without a non-streaming method.

Reply:
Good point.